### ISSUE TITLE
[apex] ApexSharingViolations - Rule doesn't flag the Apex classes declared with "without sharing"

### ISSUE THREAD

**Issue by daltarac** [pmd/pmd#5074]
Affects PMD Version: 7.2.0 / All versions

Rule: ApexSharingViolations

Please provide the rule name and a link to the rule documentation:
ApexSharingViolations (https://pmd.github.io/pmd-6.55.0/pmd_rules_apex_security.html#apexsharingviolations)

Description:
ApexSharingViolations flags the Apex classes that are not declared "with sharing". If the class is declared with "without sharing", the class is not flagged but it should be.

Code Sample demonstrating the issue:
public without sharing class ScanTestClass {
    
```
    @AuraEnabled
    public static List<Account> getAccounts()
    {
        List<Account> accountList;

        for(Integer i = 0; i < 1000; i++)
            accountList = [SELECT id, name FROM Account LIMIT 1];

        return accountList;
    }

    public void test1(String t1) {
        Database.query('SELECT Id FROM Account' + t1);
    }
}
```

Expected outcome:
PMD should report a violation at line 1, but doesn't. This is a false-negative.

Running PMD through:
Salesforce Code Analyzer

**Comment by adangel** [comment]
Thanks for the report. I'm not sure, this is a false-negative, see the rule description:

> Detect classes declared without explicit sharing mode if DML methods are used. This forces the developer to take access restrictions into account before modifying objects.

The goal of this rule seems to be to force the developer to think about sharing and make a decision - whether this is with or without, doesn't matter. As long as it is defined.

This also matches the current violation message:

> Apex classes should declare a sharing model if DML or SOQL/SOSL is used

For reference: https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_keywords_sharing.htm

It seems, this rule just implements the best practices recommendation ("Apex without an explicit sharing declaration is insecure by default. We strongly recommend that you always specify a sharing declaration for a class."). They however say, that "without sharing" should be used with caution.

Is there anything we can change, e.g. rule name, rule description, rule message to better explain what this rule does? Do we need another rule (e.g. in best practices) to suggest to always use "with sharing"?

**Comment by daltarac** [comment]
Thanks for clarifying. I agree that this is aligned with what the ApexSharingViolations rule currently does: "The goal of this rule seems to be to force the developer to think about sharing and make a decision - whether this is with or without, doesn't matter. As long as it is defined."

Would it be possible to update the documentation with something along those lines just to add more details on what the rule does?

The documentation also has the example of the code that will not be flagged by the ApexSharingViolations rule: 

`public without sharing class Foo {
    // DML operation here
}`

The example should be changed to:

`public class Foo {
    // DML operation here
}`

https://docs.pmd-code.org/latest/pmd_rules_apex_security.html#apexsharingviolations

It would be useful to have an additional rule in the security section that flags classes where:
- Record sharing is done in system mode (no declaration or "without sharing" declaration)
- DML operations are used
- Methods can be called from the front-end (@AuraEnabled, @RemoteAction and @InvocableMethod Apex method annotation) https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_annotation.htm

**Comment by dschach** [comment]
It is true that the provided example should NOT be a violation. The rule itself is written well - it requires an explicit sharing model if DML or a query is present.

It is a good feature request to require a sharing model if any method has an annotation for the fornt-end, plus any annotation starting with @http... or @rest...

I don't understand
> Record sharing is done in system mode (no declaration or "without sharing" declaration)
I think that's exactly what this rule does. And while I have a custom rule that flags all "without sharing" declarations for review, I don't think it is PMD's job to flag every "without sharing" declaration.

**Comment by dschach** [comment]
I also don't think that the OP provided code example should be a violation. The class is declared `without sharing` so PMD should not flag that, IMO.
